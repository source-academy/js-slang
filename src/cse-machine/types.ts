import * as es from 'estree'

import { Environment } from '../types'

export enum InstrType {
  RESET = 'Reset',
  WHILE = 'While',
  FOR = 'For',
  ASSIGNMENT = 'Assignment',
  UNARY_OP = 'UnaryOperation',
  BINARY_OP = 'BinaryOperation',
  POP = 'Pop',
  APPLICATION = 'Application',
  BRANCH = 'Branch',
  ENVIRONMENT = 'Environment',
  ARRAY_LITERAL = 'ArrayLiteral',
  ARRAY_ACCESS = 'ArrayAccess',
  ARRAY_ASSIGNMENT = 'ArrayAssignment',
  ARRAY_LENGTH = 'ArrayLength',
  MARKER = 'Marker',
  CONTINUE = 'Continue',
  CONTINUE_MARKER = 'ContinueMarker',
  BREAK = 'Break',
  BREAK_MARKER = 'BreakMarker'
}

interface BaseInstr {
  instrType: InstrType
  srcNode: es.Node
}

export interface WhileInstr extends BaseInstr {
  test: es.Expression
  body: es.Statement
}

export interface ForInstr extends BaseInstr {
  init: es.VariableDeclaration | es.Expression
  test: es.Expression
  update: es.Expression
  body: es.Statement
}

export interface AssmtInstr extends BaseInstr {
  symbol: string
  constant: boolean
  declaration: boolean
}

export interface UnOpInstr extends BaseInstr {
  symbol: es.UnaryOperator
}

export interface BinOpInstr extends BaseInstr {
  symbol: es.BinaryOperator
}

export interface AppInstr extends BaseInstr {
  numOfArgs: number
  srcNode: es.CallExpression
}

export interface BranchInstr extends BaseInstr {
  consequent: es.Expression | es.Statement
  alternate: es.Expression | es.Statement | null | undefined
}

export interface EnvInstr extends BaseInstr {
  env: Environment
}

export interface ArrLitInstr extends BaseInstr {
  arity: number
}

export type Instr =
  | BaseInstr
  | WhileInstr
  | AssmtInstr
  | AppInstr
  | BranchInstr
  | EnvInstr
  | ArrLitInstr

export type AgendaItem = es.Node | Instr

// Special class that cannot be found on the stash so is safe to be used
// as an indicator of a breakpoint from running the ECE machine
export class ECEBreak {}

// Special value that cannot be found on the stash so is safe to be used
// as an indicator of an error from running the ECE machine
export class ECError {
  constructor(public readonly error: any) {}
}
