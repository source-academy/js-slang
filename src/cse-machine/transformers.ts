import type { Transformer } from './patterns';

/**
 * The T component is a dictionary of mappings from syntax names to
 * their corresponding syntax rule transformers (patterns).
 *
 * Similar to the E component, there is a matching
 * "T" environment tree that is used to store the transformers.
 * as such, we need to track the transformers and update them with the environment.
 */

export class Transformers {
  private parent: Transformers | null;
  private items: Map<string, Transformer[]>;
  public constructor(parent?: Transformers) {
    this.parent = parent || null;
    this.items = new Map<string, Transformer[]>();
  }

  // only call this if you are sure that the pattern exists.
  public getPattern(name: string): Transformer[] {
    // check if the pattern exists in the current transformer
    if (this.items.has(name)) {
      return this.items.get(name) as Transformer[];
    }
    // else check if the pattern exists in the parent transformer
    if (this.parent) {
      return this.parent.getPattern(name);
    }
    // should not get here. use this properly.
    throw new Error(`Pattern ${name} not found in transformers`);
  }

  public hasPattern(name: string): boolean {
    // check if the pattern exists in the current transformer
    if (this.items.has(name)) {
      return true;
    }
    // else check if the pattern exists in the parent transformer
    if (this.parent) {
      return this.parent.hasPattern(name);
    }
    return false;
  }

  public addPattern(name: string, item: Transformer[]): void {
    this.items.set(name, item);
  }
}
