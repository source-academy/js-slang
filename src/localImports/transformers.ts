import es from 'estree'

import { ancestor } from '../utils/walkers'
import {
  createFunctionDeclaration,
  createListCallExpression,
  createLiteral,
  createPairCallExpression,
  createReturnStatement
} from './constructors'
import { isDeclaration, isDirective, isModuleDeclaration, isStatement } from './typeGuards'

const getIdentifier = (node: es.Declaration): es.Identifier | null => {
  switch (node.type) {
    case 'FunctionDeclaration':
      if (node.id === null) {
        throw new Error(
          'Encountered a FunctionDeclaration node without an identifier. This should have been caught when parsing.'
        )
      }
      return node.id
    case 'VariableDeclaration':
      const id = node.declarations[0].id
      // In Source, variable names are Identifiers.
      if (id.type !== 'Identifier') {
        throw new Error(`Expected variable name to be an Identifier, but was ${id.type} instead.`)
      }
      return id
    case 'ClassDeclaration':
      throw new Error('Exporting of class is not supported.')
  }
}

const getExportedNameToIdentifierMap = (
  nodes: es.ModuleDeclaration[]
): Record<string, es.Identifier> => {
  const exportedNameToIdentifierMap: Record<string, es.Identifier> = {}
  nodes.forEach((node: es.ModuleDeclaration): void => {
    // Only ExportNamedDeclaration nodes specify exported names.
    if (node.type !== 'ExportNamedDeclaration') {
      return
    }
    if (node.declaration) {
      const identifier = getIdentifier(node.declaration)
      if (identifier === null) {
        return
      }
      // When an ExportNamedDeclaration node has a declaration, the
      // identifier is the same as the exported name (i.e., no renaming).
      const exportedName = identifier.name
      exportedNameToIdentifierMap[exportedName] = identifier
    } else {
      // When an ExportNamedDeclaration node does not have a declaration,
      // it contains a list of names to export, i.e., export { a, b as c, d };.
      // Exported names can be renamed using the 'as' keyword. As such, the
      // exported names and their corresponding identifiers might be different.
      node.specifiers.forEach((node: es.ExportSpecifier): void => {
        const exportedName = node.exported.name
        const identifier = node.local
        exportedNameToIdentifierMap[exportedName] = identifier
      })
    }
  })
  return exportedNameToIdentifierMap
}

const getDefaultExportExpression = (
  nodes: es.ModuleDeclaration[],
  exportedNameToIdentifierMap: Partial<Record<string, es.Identifier>>
): es.Expression | null => {
  let defaultExport: es.Expression | null = null

  // Handle default exports which are parsed as ExportNamedDeclaration AST nodes.
  // 'export { name as default };' is equivalent to 'export default name;' but
  // is represented by an ExportNamedDeclaration node instead of an
  // ExportedDefaultDeclaration node.
  //
  // NOTE: If there is a named export representing the default export, its entry
  // in the map must be removed to prevent it from being treated as a named export.
  if (exportedNameToIdentifierMap['default'] !== undefined) {
    defaultExport = exportedNameToIdentifierMap['default']
    delete exportedNameToIdentifierMap['default']
  }

  nodes.forEach((node: es.ModuleDeclaration): void => {
    // Only ExportDefaultDeclaration nodes specify the default export.
    if (node.type !== 'ExportDefaultDeclaration') {
      return
    }
    if (defaultExport !== null) {
      // This should never occur because multiple default exports should have
      // been caught by the Acorn parser when parsing into an AST.
      throw new Error('Encountered multiple default exports!')
    }
    if (isDeclaration(node.declaration)) {
      const identifier = getIdentifier(node.declaration)
      if (identifier === null) {
        return
      }
      // When an ExportDefaultDeclaration node has a declaration, the
      // identifier is the same as the exported name (i.e., no renaming).
      defaultExport = identifier
    } else {
      // When an ExportDefaultDeclaration node does not have a declaration,
      // it has an expression.
      defaultExport = node.declaration
    }
  })
  return defaultExport
}

const createReturnListArguments = (
  exportedNameToIdentifierMap: Record<string, es.Identifier>
): Array<es.Expression | es.SpreadElement> => {
  return Object.entries(exportedNameToIdentifierMap).map(
    ([exportedName, identifier]: [string, es.Identifier]) => {
      const head = createLiteral(exportedName)
      const tail = identifier
      return createPairCallExpression(head, tail)
    }
  )
}

const removeDirectives = (
  nodes: Array<es.Directive | es.Statement | es.ModuleDeclaration>
): Array<es.Statement | es.ModuleDeclaration> => {
  return nodes.filter(
    (
      node: es.Directive | es.Statement | es.ModuleDeclaration
    ): node is es.Statement | es.ModuleDeclaration => !isDirective(node)
  )
}

const removeModuleDeclarations = (
  nodes: Array<es.Statement | es.ModuleDeclaration>
): es.Statement[] => {
  const statements: es.Statement[] = []
  nodes.forEach((node: es.Statement | es.ModuleDeclaration): void => {
    if (isStatement(node)) {
      statements.push(node)
      return
    }
    switch (node.type) {
      case 'ExportNamedDeclaration':
        if (node.declaration) {
          statements.push(node.declaration)
        }
        break
      case 'ExportDefaultDeclaration':
        if (isDeclaration(node.declaration)) {
          statements.push(node.declaration)
        }
        break
      case 'ExportAllDeclaration':
        throw new Error('Not implemented yet.')
    }
  })
  return statements
}

export const transformImportedFile = (
  program: es.Program,
  iifeIdentifier: string
): es.FunctionDeclaration => {
  const moduleDeclarations = program.body.filter(isModuleDeclaration)
  const exportedNameToIdentifierMap = getExportedNameToIdentifierMap(moduleDeclarations)
  const defaultExportExpression = getDefaultExportExpression(
    moduleDeclarations,
    exportedNameToIdentifierMap
  )

  const defaultExport = defaultExportExpression ?? createLiteral(null)
  const namedExports = createListCallExpression(
    createReturnListArguments(exportedNameToIdentifierMap)
  )
  const returnStatement = createReturnStatement(
    createPairCallExpression(defaultExport, namedExports)
  )

  const programStatements = removeModuleDeclarations(removeDirectives(program.body))
  const iifeBody = [...programStatements, returnStatement]

  return createFunctionDeclaration(iifeIdentifier, [], iifeBody)
}

/**
 * Removes all export-related nodes from the AST.
 *
 * Export-related AST nodes are only needed in the local imports pre-processing
 * step to determine which functions/variables/expressions should be made
 * available to other files/modules. After which, they have no functional effect
 * on program evaluation.
 *
 * @param program The AST which should be stripped of export-related nodes.
 */
export const removeExports = (program: es.Program): void => {
  ancestor(program, {
    // TODO: Handle other export AST nodes.
    ExportNamedDeclaration(node: es.ExportNamedDeclaration, ancestors: es.Node[]) {
      // The ancestors array contains the current node, meaning that the
      // parent node is the second last node of the array.
      const parent = ancestors[ancestors.length - 2]
      // The parent node of an ExportNamedDeclaration node must be a Program node.
      if (parent.type !== 'Program') {
        return
      }
      const nodeIndex = parent.body.findIndex(n => n === node)
      if (node.declaration) {
        // If the ExportNamedDeclaration node contains a declaration, replace
        // it with the declaration node in its parent node's body.
        parent.body[nodeIndex] = node.declaration
      } else {
        // Otherwise, remove the ExportNamedDeclaration node in its parent node's body.
        parent.body.splice(nodeIndex, 1)
      }
    },
    ExportDefaultDeclaration(node: es.ExportDefaultDeclaration, ancestors: es.Node[]) {
      // The ancestors array contains the current node, meaning that the
      // parent node is the second last node of the array.
      const parent = ancestors[ancestors.length - 2]
      // The parent node of an ExportNamedDeclaration node must be a Program node.
      if (parent.type !== 'Program') {
        return
      }
      const nodeIndex = parent.body.findIndex(n => n === node)
      // 'node.declaration' can be either a Declaration or an Expression.
      if (isDeclaration(node.declaration)) {
        // If the ExportDefaultDeclaration node contains a declaration, replace
        // it with the declaration node in its parent node's body.
        parent.body[nodeIndex] = node.declaration
      } else {
        // Otherwise, the ExportDefaultDeclaration node contains a statement.
        // Remove the ExportDefaultDeclaration node in its parent node's body.
        parent.body.splice(nodeIndex, 1)
      }
    }
  })
}

/**
 * Returns whether a string is a file path. We define a file
 * path to be any string containing the '/' character.
 *
 * @param value The value of the string.
 */
const isFilePath = (value: string): boolean => {
  return value.includes('/')
}

/**
 * Returns whether a string is a reference to a Source module.
 * We define a Source module name to be any string that is not
 * a file path.
 *
 * Source module import:           `import { x } from "module";`
 * Local (relative) module import: `import { x } from "./module";`
 * Local (absolute) module import: `import { x } from "/dir/dir2/module";`
 *
 * @param value The value of the string.
 */
export const isSourceModule = (value: string): boolean => {
  return !isFilePath(value)
}

export const removeNonSourceModuleImports = (program: es.Program): void => {
  // First pass: remove all import AST nodes which are unused by Source modules.
  ancestor(program, {
    ImportSpecifier(_node: es.ImportSpecifier, _ancestors: es.Node[]): void {
      // Nothing to do here since ImportSpecifier nodes are used by Source modules.
    },
    ImportDefaultSpecifier(node: es.ImportDefaultSpecifier, ancestors: es.Node[]): void {
      // The ancestors array contains the current node, meaning that the
      // parent node is the second last node of the array.
      const parent = ancestors[ancestors.length - 2]
      // The parent node of an ImportDefaultSpecifier node must be an ImportDeclaration node.
      if (parent.type !== 'ImportDeclaration') {
        return
      }
      const nodeIndex = parent.specifiers.findIndex(n => n === node)
      // Remove the ImportDefaultSpecifier node in its parent node's array of specifiers.
      // This is because Source modules do not support default imports.
      parent.specifiers.splice(nodeIndex, 1)
    },
    ImportNamespaceSpecifier(node: es.ImportNamespaceSpecifier, ancestors: es.Node[]): void {
      // The ancestors array contains the current node, meaning that the
      // parent node is the second last node of the array.
      const parent = ancestors[ancestors.length - 2]
      // The parent node of an ImportNamespaceSpecifier node must be an ImportDeclaration node.
      if (parent.type !== 'ImportDeclaration') {
        return
      }
      const nodeIndex = parent.specifiers.findIndex(n => n === node)
      // Remove the ImportNamespaceSpecifier node in its parent node's array of specifiers.
      // This is because Source modules do not support namespace imports.
      parent.specifiers.splice(nodeIndex, 1)
    }
  })

  const removeImportDeclaration = (node: es.ImportDeclaration, ancestors: es.Node[]): void => {
    // The ancestors array contains the current node, meaning that the
    // parent node is the second last node of the array.
    const parent = ancestors[ancestors.length - 2]
    // The parent node of an ImportDeclaration node must be a Program node.
    if (parent.type !== 'Program') {
      return
    }
    const nodeIndex = parent.body.findIndex(n => n === node)
    // Remove the ImportDeclaration node in its parent node's body.
    parent.body.splice(nodeIndex, 1)
  }
  // Second pass: remove all ImportDeclaration nodes for non-Source modules, or that do not
  // have any specifiers (thus being functionally useless).
  ancestor(program, {
    ImportDeclaration(node: es.ImportDeclaration, ancestors: es.Node[]): void {
      // Module names must always be strings in Source.
      if (typeof node.source.value !== 'string') {
        return
      }
      // ImportDeclaration nodes without any specifiers are functionally useless and are thus removed.
      if (node.specifiers.length === 0) {
        removeImportDeclaration(node, ancestors)
        return
      }
      // Non-Source modules should already have been handled in the pre-processing step and are no
      // longer needed. They must be removed to avoid being treated as Source modules.
      if (!isSourceModule(node.source.value)) {
        removeImportDeclaration(node, ancestors)
      }
    }
  })
}
