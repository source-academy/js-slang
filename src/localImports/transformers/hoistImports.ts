import es from 'estree'

import { isImportDeclaration } from '../typeGuards'

export const hoistImports = (program: es.Program): void => {
  const importDeclarations = program.body.filter(isImportDeclaration)
  const nonImportDeclarations = program.body.filter(
    (node: es.Directive | es.Statement | es.ModuleDeclaration): boolean =>
      !isImportDeclaration(node)
  )
  program.body = [...importDeclarations, ...nonImportDeclarations]
}
