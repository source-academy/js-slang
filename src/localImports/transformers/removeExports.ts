import es from 'estree'

import { ancestor } from '../../utils/walkers'
import { isDeclaration } from '../typeGuards'

/**
 * Removes all export-related nodes from the AST.
 *
 * Export-related AST nodes are only needed in the local imports pre-processing
 * step to determine which functions/variables/expressions should be made
 * available to other files/modules. After which, they have no functional effect
 * on program evaluation.
 *
 * @param program The AST which should be stripped of export-related nodes.
 */
export const removeExports = (program: es.Program): void => {
  ancestor(program, {
    // TODO: Handle other export AST nodes.
    ExportNamedDeclaration(
      node: es.ExportNamedDeclaration,
      _state: es.Node[],
      ancestors: es.Node[]
    ) {
      // The ancestors array contains the current node, meaning that the
      // parent node is the second last node of the array.
      const parent = ancestors[ancestors.length - 2]
      // The parent node of an ExportNamedDeclaration node must be a Program node.
      if (parent.type !== 'Program') {
        return
      }
      const nodeIndex = parent.body.findIndex(n => n === node)
      if (node.declaration) {
        // If the ExportNamedDeclaration node contains a declaration, replace
        // it with the declaration node in its parent node's body.
        parent.body[nodeIndex] = node.declaration
      } else {
        // Otherwise, remove the ExportNamedDeclaration node in its parent node's body.
        parent.body.splice(nodeIndex, 1)
      }
    },
    ExportDefaultDeclaration(
      node: es.ExportDefaultDeclaration,
      _state: es.Node[],
      ancestors: es.Node[]
    ) {
      // The ancestors array contains the current node, meaning that the
      // parent node is the second last node of the array.
      const parent = ancestors[ancestors.length - 2]
      // The parent node of an ExportNamedDeclaration node must be a Program node.
      if (parent.type !== 'Program') {
        return
      }
      const nodeIndex = parent.body.findIndex(n => n === node)
      // 'node.declaration' can be either a Declaration or an Expression.
      if (isDeclaration(node.declaration)) {
        // If the ExportDefaultDeclaration node contains a declaration, replace
        // it with the declaration node in its parent node's body.
        parent.body[nodeIndex] = node.declaration
      } else {
        // Otherwise, the ExportDefaultDeclaration node contains a statement.
        // Remove the ExportDefaultDeclaration node in its parent node's body.
        parent.body.splice(nodeIndex, 1)
      }
    }
  })
}
