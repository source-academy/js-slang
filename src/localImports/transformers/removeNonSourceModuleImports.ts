import es from 'estree'

import { ancestor } from '../../utils/walkers'
import { isFilePath } from '../filePaths'

/**
 * Returns whether a module name refers to a Source module.
 * We define a Source module name to be any string that is not
 * a file path.
 *
 * Source module import:           `import { x } from "module";`
 * Local (relative) module import: `import { x } from "./module";`
 * Local (absolute) module import: `import { x } from "/dir/dir2/module";`
 *
 * @param moduleName The name of the module.
 */
export const isSourceModule = (moduleName: string): boolean => {
  return !isFilePath(moduleName)
}

/**
 * Removes all non-Source module import-related nodes from the AST.
 *
 * All import-related nodes which are not removed in the pre-processing
 * step will be treated by the Source modules loader as a Source module.
 * If a Source module by the same name does not exist, the program
 * evaluation will error out. As such, this function removes all
 * import-related AST nodes which the Source module loader does not
 * support, as well as ImportDeclaration nodes for local module imports.
 *
 * The definition of whether a module is a local module or a Source
 * module depends on the implementation of the `isSourceModule` function.
 *
 * @param program The AST which should be stripped of non-Source module
 *                import-related nodes.
 */
export const removeNonSourceModuleImports = (program: es.Program): void => {
  // First pass: remove all import AST nodes which are unused by Source modules.
  ancestor(program, {
    ImportSpecifier(_node: es.ImportSpecifier, _state: es.Node[], _ancestors: es.Node[]): void {
      // Nothing to do here since ImportSpecifier nodes are used by Source modules.
    },
    ImportDefaultSpecifier(
      node: es.ImportDefaultSpecifier,
      _state: es.Node[],
      ancestors: es.Node[]
    ): void {
      // The ancestors array contains the current node, meaning that the
      // parent node is the second last node of the array.
      const parent = ancestors[ancestors.length - 2]
      // The parent node of an ImportDefaultSpecifier node must be an ImportDeclaration node.
      if (parent.type !== 'ImportDeclaration') {
        return
      }
      const nodeIndex = parent.specifiers.findIndex(n => n === node)
      // Remove the ImportDefaultSpecifier node in its parent node's array of specifiers.
      // This is because Source modules do not support default imports.
      parent.specifiers.splice(nodeIndex, 1)
    },
    ImportNamespaceSpecifier(
      node: es.ImportNamespaceSpecifier,
      _state: es.Node[],
      ancestors: es.Node[]
    ): void {
      // The ancestors array contains the current node, meaning that the
      // parent node is the second last node of the array.
      const parent = ancestors[ancestors.length - 2]
      // The parent node of an ImportNamespaceSpecifier node must be an ImportDeclaration node.
      if (parent.type !== 'ImportDeclaration') {
        return
      }
      const nodeIndex = parent.specifiers.findIndex(n => n === node)
      // Remove the ImportNamespaceSpecifier node in its parent node's array of specifiers.
      // This is because Source modules do not support namespace imports.
      parent.specifiers.splice(nodeIndex, 1)
    }
  })

  // Operate on a copy of the Program node's body to prevent the walk from missing ImportDeclaration nodes.
  const programBody = [...program.body]
  const removeImportDeclaration = (node: es.ImportDeclaration, ancestors: es.Node[]): void => {
    // The ancestors array contains the current node, meaning that the
    // parent node is the second last node of the array.
    const parent = ancestors[ancestors.length - 2]
    // The parent node of an ImportDeclaration node must be a Program node.
    if (parent.type !== 'Program') {
      return
    }
    const nodeIndex = programBody.findIndex(n => n === node)
    // Remove the ImportDeclaration node in its parent node's body.
    programBody.splice(nodeIndex, 1)
  }
  // Second pass: remove all ImportDeclaration nodes for non-Source modules, or that do not
  // have any specifiers (thus being functionally useless).
  ancestor(program, {
    ImportDeclaration(node: es.ImportDeclaration, _state: es.Node[], ancestors: es.Node[]): void {
      if (typeof node.source.value !== 'string') {
        throw new Error('Module names must be strings.')
      }
      // ImportDeclaration nodes without any specifiers are functionally useless and are thus removed.
      if (node.specifiers.length === 0) {
        removeImportDeclaration(node, ancestors)
        return
      }
      // Non-Source modules should already have been handled in the pre-processing step and are no
      // longer needed. They must be removed to avoid being treated as Source modules.
      if (!isSourceModule(node.source.value)) {
        removeImportDeclaration(node, ancestors)
      }
    }
  })
  program.body = programBody
}
