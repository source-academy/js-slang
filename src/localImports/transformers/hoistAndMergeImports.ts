import es from 'estree'
import * as _ from 'lodash'

import { createImportDeclaration, createLiteral } from '../constructors/baseConstructors'
import { cloneAndStripImportSpecifier } from '../constructors/contextSpecificConstructors'
import { isImportDeclaration } from '../typeGuards'
import { isSourceModule } from './removeNonSourceModuleImports'

/**
 * Hoists import declarations to the top of the program & merges duplicate
 * imports for the same module.
 *
 * Note that two modules are the same if and only if their import source
 * is the same. This function does not resolve paths against a base
 * directory. If such a functionality is required, this function will
 * need to be modified.
 *
 * @param program The AST which should have its ImportDeclaration nodes
 *                hoisted & duplicate imports merged.
 */
export const hoistAndMergeImports = (program: es.Program): void => {
  // Separate import declarations from non-import declarations.
  const importDeclarations = program.body.filter(isImportDeclaration)
  const nonImportDeclarations = program.body.filter(
    (node: es.Directive | es.Statement | es.ModuleDeclaration): boolean =>
      !isImportDeclaration(node)
  )

  // Merge import sources & specifiers.
  const importSourceToSpecifiersMap: Map<
    string,
    Array<es.ImportSpecifier | es.ImportDefaultSpecifier | es.ImportNamespaceSpecifier>
  > = new Map()
  for (const importDeclaration of importDeclarations) {
    const importSource = importDeclaration.source.value
    if (typeof importSource !== 'string') {
      throw new Error('Module names must be strings.')
    }
    const specifiers = importSourceToSpecifiersMap.get(importSource) ?? []
    for (const specifier of importDeclaration.specifiers) {
      // The Acorn parser adds extra information to AST nodes that are not
      // part of the ESTree types. As such, we need to clone and strip
      // the import specifier AST nodes to get a canonical representation
      // that we can use to keep track of whether the import specifier
      // is a duplicate or not.
      const strippedSpecifier = cloneAndStripImportSpecifier(specifier)
      // Note that we cannot make use of JavaScript's built-in Set class
      // as it compares references for objects.
      const isSpecifierDuplicate =
        specifiers.filter(
          (
            specifier: es.ImportSpecifier | es.ImportDefaultSpecifier | es.ImportNamespaceSpecifier
          ): boolean => {
            return _.isEqual(strippedSpecifier, specifier)
          }
        ).length !== 0
      if (isSpecifierDuplicate) {
        continue
      }
      specifiers.push(strippedSpecifier)
    }
    importSourceToSpecifiersMap.set(importSource, specifiers)
  }

  // Convert the merged import sources & specifiers back into import declarations.
  const mergedImportDeclarations: es.ImportDeclaration[] = []
  importSourceToSpecifiersMap.forEach(
    (
      specifiers: Array<
        es.ImportSpecifier | es.ImportDefaultSpecifier | es.ImportNamespaceSpecifier
      >,
      importSource: string
    ): void => {
      mergedImportDeclarations.push(
        createImportDeclaration(specifiers, createLiteral(importSource))
      )
    }
  )

  // Hoist the merged import declarations to the top of the program body.
  program.body = [...mergedImportDeclarations, ...nonImportDeclarations]
}

export default function hoistAndMergeImportsNew(programs: es.Program[]) {
  const allNodes = programs.flatMap(({ body }) => body)

  const importNodes = allNodes.filter(
    (node): node is es.ImportDeclaration => node.type === 'ImportDeclaration'
  )
  const importsToSpecifiers = new Map<string, Map<string, Set<string>>>()
  for (const node of importNodes) {
    const source = node.source!.value as string
    // We no longer need imports from non-source modules, so we can just ignore them
    if (!isSourceModule(source)) continue

    if (!importsToSpecifiers.has(source)) {
      importsToSpecifiers.set(source, new Map())
    }

    const specifierMap = importsToSpecifiers.get(source)!
    node.specifiers.forEach(spec => {
      let importingName: string
      switch (spec.type) {
        case 'ImportSpecifier': {
          importingName = spec.imported.name
          break
        }
        case 'ImportDefaultSpecifier': {
          importingName = 'default'
          break
        }
        case 'ImportNamespaceSpecifier': {
          // TODO handle
          throw new Error()
        }
      }

      if (!specifierMap.has(importingName)) {
        specifierMap.set(importingName, new Set())
      }
      specifierMap.get(importingName)!.add(spec.local.name)
    })
  }

  // Every distinct source module being imported is given its own ImportDeclaration node
  const importDeclarations = Array.from(importsToSpecifiers.entries()).map(
    ([moduleName, imports]) => {
      // Across different modules, the user may choose to alias some of the declarations, so we keep track,
      // of all the different aliases used for each unique imported symbol
      const specifiers = Array.from(imports.entries()).flatMap(([importedName, aliases]) => {
        if (importedName === 'default') {
          return Array.from(aliases).map(
            alias =>
              ({
                type: 'ImportDefaultSpecifier',
                local: {
                  type: 'Identifier',
                  name: alias
                }
              } as es.ImportDefaultSpecifier)
          ) as (es.ImportSpecifier | es.ImportDefaultSpecifier)[]
        } else {
          return Array.from(aliases).map(
            alias =>
              ({
                type: 'ImportSpecifier',
                imported: {
                  type: 'Identifier',
                  name: importedName
                },
                local: {
                  type: 'Identifier',
                  name: alias
                }
              } as es.ImportSpecifier)
          )
        }
      })

      const decl: es.ImportDeclaration = {
        type: 'ImportDeclaration',
        source: {
          type: 'Literal',
          value: moduleName
        },
        specifiers
      }
      return decl
    }
  )
  return importDeclarations
}
