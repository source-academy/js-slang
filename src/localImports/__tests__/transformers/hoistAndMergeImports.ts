import { mockContext } from '../../../mocks/context'
import { parse } from '../../../parser/parser'
import { Chapter } from '../../../types'
import { hoistAndMergeImports } from '../../transformers/hoistAndMergeImports'
import { parseCodeError, stripLocationInfo } from '../utils'

describe('hoistAndMergeImports', () => {
  let actualContext = mockContext(Chapter.LIBRARY_PARSER)
  let expectedContext = mockContext(Chapter.LIBRARY_PARSER)

  beforeEach(() => {
    actualContext = mockContext(Chapter.LIBRARY_PARSER)
    expectedContext = mockContext(Chapter.LIBRARY_PARSER)
  })

  const assertASTsAreEquivalent = (actualCode: string, expectedCode: string): void => {
    const actualProgram = parse(actualCode, actualContext)
    const expectedProgram = parse(expectedCode, expectedContext)
    if (actualProgram === null || expectedProgram === null) {
      throw parseCodeError
    }

    hoistAndMergeImports(actualProgram)
    expect(stripLocationInfo(actualProgram)).toEqual(stripLocationInfo(expectedProgram))
  }

  test('hoists import declarations to the top of the program', () => {
    const actualCode = `
      function square(x) {
        return x * x;
      }

      import { a, b, c } from "./a.js";

      export { square };

      import x from "source-module";

      square(3);
    `
    const expectedCode = `
      import { a, b, c } from "./a.js";
      import x from "source-module";

      function square(x) {
        return x * x;
      }

      export { square };

      square(3);
    `
    assertASTsAreEquivalent(actualCode, expectedCode)
  })

  test('merges import declarations from the same module', () => {
    const actualCode = `
      import { a, b, c } from "./a.js";
      import { d } from "./a.js";
      import { x } from "./b.js";
      import { e, f } from "./a.js";
    `
    const expectedCode = `
      import { a, b, c, d, e, f } from "./a.js";
      import { x } from "./b.js";
    `
    assertASTsAreEquivalent(actualCode, expectedCode)
  })
})
