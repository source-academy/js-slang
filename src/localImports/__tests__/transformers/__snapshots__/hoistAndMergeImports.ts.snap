// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`hoistAndMergeImports hoists import declarations to the top of the program 1`] = `
"import x from \\"source-module\\";
function square(x) {
  return x * x;
}
import {a, b, c} from \\"./a.js\\";
export {square};
import x from \\"source-module\\";
square(3);
"
`;

exports[`hoistAndMergeImports merges import declarations from the same module 1`] = `
"import {a, b, c} from \\"./a.js\\";
import {d} from \\"./a.js\\";
import {x} from \\"./b.js\\";
import {e, f} from \\"./a.js\\";
"
`;
